### ISSUE TITLE
Improper quoting in IMAP KEYWORD search

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4392]
|[<img alt="jothan's avatar" src="https://avatars.githubusercontent.com/u/193014?s=50" width="50" height="50">](https://github.com/jothan)| @jothan reported|
|-|-|
|Trac ID|trac#4392|
|Type|defect|
|Created|2010-04-07 15:16:14Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-keyword-atoms-4392-2|

Cyrus IMAP refuses keyword queries generated by Twisted with an "invalid flag name error". Looking in the IMAP RFC, keyword flags are "atoms" in the grammar. That means it should not be quoted.

IMAP4Server.atomre could be used to make sure no silly caracters are injected inside the IMAP query.

Reference: http://tools.ietf.org/html/rfc3501#section-9
Look for "flag-keyword"

Test case:
```
from twisted.mail.imap4 import Query

print Query(keyword='learnasspam')
```

Outputs '(KEYWORD "learnasspam")'
Should be: '(KEYWORD learnasspam)'

This is present in Twisted 10 and the lastest SVN.
Patch attached.

Attachments:

* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82e/82e9eac8bcb19566d9204e9d0a99b6965d44b9bc/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (335 bytes) - added by jothan on 2010-04-07 15:17:08Z - Patch to fix this bug.
* [4392-IMAPKeyword-20120629.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82e/82e9eac8bcb19566d9204e9d0a99b6965d44b9bc/41408114e19e496f7d37bac2d9117e77ae8015c2.patch) (1580 bytes) - added by argonemyth on 2012-06-29 16:23:42Z - 
* [4392-IMAPKeyword-20120711.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82e/82e9eac8bcb19566d9204e9d0a99b6965d44b9bc/38c81461625430fce5e3c28efef6b9ebb4af8265.patch) (3724 bytes) - added by argonemyth on 2012-07-11 19:11:56Z - Second try to fix the bug!
* [4392-IMAPKeywordFlag-NoQuotes-20120730.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82e/82e9eac8bcb19566d9204e9d0a99b6965d44b9bc/9f16dda104f3f197ed7d2037a029cc06f3d86467.patch) (1661 bytes) - added by argonemyth on 2012-07-30 20:37:39Z - Adds '''KEYWORD''' and '''UNKEYWORD''' to _NO_QUOTES
* [4392-IMAPKeywordFlag-StripAtom-20120730.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82e/82e9eac8bcb19566d9204e9d0a99b6965d44b9bc/eebec052e69f415aae411dd668f11cc78e490178.patch) (3088 bytes) - added by argonemyth on 2012-07-30 20:38:06Z - Strips the invalid characters

<details><summary>Searchable metadata</summary>

```
trac-id__4392 4392
type__defect defect
reporter__jothan jothan
priority__normal normal
milestone__ 
branch__branches_imap4_keyword_atoms_4392_2 branches/imap4-keyword-atoms-4392-2
branch_author__argonemyth__exarkun argonemyth, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__imap_keyword imap keyword
time__1270653374000000 1270653374000000
changetime__1358987794000000 1358987794000000
version__None None
owner__exarkun exarkun
cc__exarkun cc__argonemyth cc__jesstess cc__danmil cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @jothan|
|-|-|

Hi jothan,

Thanks very much for tracking this down and contributing a patch.  I wonder if you could do a little bit more work, and add unit tests for this change?  Also, I think you're right that the keyword value should be checked and invalid keywords rejected with an exception.  Do you feel up to implementing that as well?

When you have a new patch, please add the "review" keyword to the ticket and change the owner to no-one (the empty string at the top of the assignee list) so another developer knows there's something new to look at.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_2"></a>|@argonemyth set owner to @argonemyth|
|-|-|

I will add unit tests for the change!

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_3"></a>|@argonemyth removed owner|
|-|-|

Hey, My first attempt to patch the source code, please be patient with me. 

From how I understand, I don't think it's necessary to raise an exception here. Even we should warn the user somehow, it should be in the implementation of mailbox, not in the Query() function or the IMAP4Server class. 

The new fix (4392-IMAPKeyword-20120629.patch) strips away all the invalid characters for an atom argument. I don't think it will change what an user intended to do, as an flag with invalid characters won't be created at the first place. In most cases, no one would query a keyword flag with invalid characters, unless a typo occurs. If you think the new fix is a bit "destructive", the original fix is simpler and better. Let me know which version you want to keep.

Also, test_queryAtomArgs() is added to IMAP4HelperTestCase.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jothan|
|-|-|

Thanks!  A few comments on the change:

  1. The docstring for `test_queryAtomArgs` doesn't follow the convention.  Instead, it should explain the behavior being tested, without using the words "Test", "correct", or "should".  For example, "When passed the C{keyword} argument, L{imap4.Query} returns a C{str} that includes only the legal I{KEYWORD} characters from the string passed to it.".  A reference to the section of the RFC that defines what "legal" means would also be great.
  1. There's some trailing whitespace in the test, too.  Please remove all spaces at the end of lines.
  1. The string inside the `if k in _ATOM_ARGS:` check is misplaced.  docstrings only belong on modules, classes, and functions.  Elsewhere, use comments.
  1. Are there other ATOM arguments besides _KEYWORD_?  If not, using a tuple and containment check here doesn't make much sense.  If so, maybe we should fix them too?
  1. The regular expression used to remove illegal characters is a bit complicated.  The test does not exercise all of it.  Can you expand the test to verify that every single illegal character is removed, and every single legal character is retained?

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @argonemyth|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_7"></a>|@argonemyth removed owner|
|-|-|

Thanks for the comments! 

>   4. Are there other ATOM arguments besides _KEYWORD_?  If not, using a tuple and containment check here doesn't make much sense.  If so, maybe we should fix them too? 


Well, according to the spec (http://tools.ietf.org/html/rfc3501#section-6.4.4), **KEYWORD** is the only search key needs an atom argument. I removed _ATOM_ARGS variable from the patch.


>   5. The regular expression used to remove illegal characters is a bit complicated.  The test does not exercise all of it.  Can you expand the test to verify that every single illegal character is removed, and every single legal character is retained?

I included all the 'illegal characters' in the test, if I missed any, let me know. 

rfc3501 doesn't seem to specify what are **illegal** characters in an atom, I just used the comments and borrowed the regular expression that's already in imap4.py.

Please review the modified patch & thanks :)

**Comment by twisted-trac** [comment]
|<img alt="danmil's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|danmil set owner to @argonemyth|
|-|-|


I'm mostly certain that this overall approach is incorrect -- specifically, I think it's a very bad idea to silently strip out any offending characters from the keyword argument 'keyword' to Query().

The Query function (it's not a class, despite the initial cap), is a tool for someone using the imap4.py *client*, programmatically.  It's not, unless I'm deeply misunderstanding it, ever used by the server code, in any way.  Thus, I think it'd be far better to explicitly raise an imap4.IllegalQueryError if any bad characters are detected (fail fast and loud).  If a developer is trying to construct a keyword search with one of those bad characters, they will be deeply confused by 'successful' searches that don't match what they're looking for.

If you want to get fancier, you can try to figure out how atoms escape those characters and automatically do that, but I think just failing loud and fast is a solid improvement over the current behavior (which, if I understand it, is broken for all normal use, at least against the Cyrus server)

Three other points:

 - There is an UNKEYWORD argument which looks like it needs to be treated similarly (see l. 4084 in imap4.py and 6.4.4 in RFC 3501)

 - However you end up validating the proper atom nature of the keyword, you should definitely try to reuse IMAP4Server.atomre, as suggested in the original ticket comment.  That's a good single point of truth for the lexical structure of an atom.  It's possible it'll be useful to then move that out of that class, but do try to not replicate it.

 - When you're constructing regex's in python the 'raw string' mode is extremely useful.  By prefixing the string literal with the letter 'r', it tells the python lexer to not touch backslashes.  E.g. the existing atomre is defined:

```python
   atomre = re.compile(r'(?P<atom>[^\](){%*"\\\x00-\x20\x80-\xff]+)( (?P<rest>.*$)|$)')
```

**Comment by twisted-trac** [comment]
|<img alt="danmil's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|danmil commented|
|-|-|

One other thing I just realized: there is no way to 'escape' bad characters.  The keyword being passed in can *only* be an atom.  So just fail loud and fast.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_11"></a>|@argonemyth commented|
|-|-|

Replying to [danmil](#note_8):

> I'm mostly certain that this overall approach is incorrect -- specifically, I think it's a very bad idea to silently strip out any offending characters from the keyword argument 'keyword' to Query().
>
> The Query function (it's not a class, despite the initial cap), is a tool for someone using the imap4.py *client*, programmatically.  It's not, unless I'm deeply misunderstanding it, ever used by the server code, in any way.  Thus, I think it'd be far better to explicitly raise an imap4.IllegalQueryError if any bad characters are detected (fail fast and loud).  If a developer is trying to construct a keyword search with one of those bad characters, they will be deeply confused by 'successful' searches that don't match what they're looking for.

Like I already said, stripping away the invalid characters was just a **suggestion**. The reason behind it is that it's not the regular 'keyword search', it's a keyword **flag** search. According the the spec, a server should not allow an user to create a flag with invalid characters.

The point of this ticket is really not about atom IMHO, it's about the quotations - jothan pointed out that the **keyword flag** should not have the quotations around because it should be an atom. 

Maybe I did more than needed here. So, someone please tell me how we should fix this ticket properly!(/me looking at exarkun!)  
 

>  - There is an UNKEYWORD argument which looks like it needs to be treated similarly (see l. 4084 in imap4.py and 6.4.4 in RFC 3501)

Thanks for pointing this out! I don't know how I missed it!! I searched the whole memo for the **flag**...good to have extra eyes to look for those details! :)

**Comment by twisted-trac** [comment]
|<img alt="danmil's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|danmil commented|
|-|-|

First off, apologies if my tone was strong -- I think it's **fantastic** that you're putting some work into cleaning up `twisted.mail` (I'm actually using imap4.py at my job, which is why I'm getting involved).

Second off, I do get that this is keyword flag search, and that according to the RFC, the server should not allow a user to create a flag with invalid characters (that's what my later comment was -- me realizing that).

I see that the point of the ticket is to get rid of the extra quote marks.  My concern is that silently removing the invalid characters **on the client side** is an unfortunate kind of silent failure.  We can only get into that situation if someone is using the imap4.py client, and then attempts to issue a malformed keyword search.  In that case, I think raising the (existing) IllegalQueryError is much more helpful to them -- either they've made a mistake, and should know right away, or they're trying to do something they can't, in which case they'll be confused by getting results which don't match what they asked for.

Having the server be liberal in what it accepts -- I could see that.  But having the programmatic client try to silently fix things up -- that's what makes me nervous.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_13"></a>|@argonemyth removed owner|
|-|-|

I am going to submit two patches:

1. 4392-IMAPKeywordFlag-NoQuotes-20120730.patch

- Simply added **KEYWORD** and **UNKEYWORD** to _NO_QUOTES.

- Added two tests.

2. 4392-IMAPKeywordFlag-StripAtom-20120730.patch

- I finally found the definition of atoms in RFC2822, according to section 3.2.1: http://tools.ietf.org/html/rfc2822#section-3.2.1, the invalid characters are not exactly the same as in **IMAP4Server.atomre**, so, I made modifications to the regular expression & tests. I didn't change IMAP4Server.atomre though, if **atomre** needs to be changed (if I am read the rfc2822 correctly), I think it's better to open another ticket, as it's not directly related to the bug in this ticket.

- Added a new test for the **UNKEYWORD** query.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [36597]) Branching to 'imap4-keyword-atoms-4392'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [36598]) Apply 4392-IMAPKeywordFlag-NoQuotes-20120730.patch and 4392-IMAPKeywordFlag-StripAtom-20120730.patch 

refs [#4392](https://github.com/twisted/twisted/issues/4392)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

  1. > I finally found the definition of atoms in RFC2822, according to section 3.2.1

This is mistaken.  RFC 2822 is an SMTP RFC, not an IMAP4 RFC.  The two protocols may coincidentally share some vocabulary, but it's only a coincidence.  The definition of atom from RFC 2822 is not applicable here.

Instead, the definition is this (from RFC 3501):

```
atom            = 1*ATOM-CHAR
ATOM-CHAR       = <any CHAR except atom-specials>
atom-specials   = "(" / ")" / "{" / SP / CTL / list-wildcards /
                  quoted-specials / resp-specials
list-wildcards  = "%" / "*"
quoted-specials = DQUOTE / "\"
resp-specials   = "]"
```

It appears as though the definition of `DQUOTE` is omitted, but I think it is supposed to be defined as `"` (iow, it is "double quote").  Likewise, `SP` is probably ` ` and `CTL` is probably the conventional definition of "ascii control characters".

I believe the upshot of all that is that `IMAP4Server.atomre` is correct.

For your consideration, though, I think that if `atomre` had not been correct (or turns out still to not be correct for some reason), the better course of action would be to fix it (quite possible as part of a separate ticket, as you suggest!) _first_ and then use its corrected form to resolve this ticket without largely duplicating its definition (particularly as these definitions both involve somewhat complicated regular expressions).

  2. _keyword_ is misspelled in most of the test method docstrings.
  3. I think the behavior of stripping invalid characters is fine.  It's not as good as raising an exception to signal to application developers they're doing something wrong, but it's an improvement over the current behavior (which is to speak the protocol incorrectly).  The new behavior will be to speak the protocol correctly, even given incorrect inputs.  A potential future improvement might be to issue a warning for incorrect inputs, and then later on the warning could be turned into an exception (at which point no one should be surprised).  It would also be nice if the set of legal bytes were exposed by this API somehow, since it is a bit difficult to construct that set given just the RFC.  These are all future improvements that belong on separate tickets, though.
  4. Assertions in loops result in difficult to understand failures.  Which iteration of the loop did the assertion fail on?  What value was it inspecting?  The `assert` methods accept an extra argument to give a more information error message; this feature is useful where loops really are helpful in a test.  Another approach is just not to loop, and that's usually probably a better idea (I know the IMAP4 tests are full of loops, that's my fault, but at the time I didn't know any better).
  5. Some docstrings still say _should_, they just need to commit to it and state some behavior without reservation (ie, just leave _should_ out).
  6. There's some commented out code in one of the test methods that should just be deleted.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [36599]) Address review feedback

Also:

  - Take KEYWORD and UNKEYWORD out of _NO_QUOTES.  It is true they do not get quotes, but other code changes are responsible for that.
  - Refactor some tests to duplicate less code
  - Refactor the definition of atomre so its definition can contribute to this fix (previously it was impossible to use atomre to implement the necessary filtering in Query)

refs [#4392](https://github.com/twisted/twisted/issues/4392)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I went ahead and addressed those points, plus a few more.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imap4-keyword-atoms-4392).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj set owner to @exarkun<br>@wallrj set status to `new`|
|-|-|

# Code Review
This branch looks fine to me.

1. All tests pass on my Fedora 17 system
2. Buildbot tests seem to have passed
3. Branch merged cleanly to trunk
4. It would have been useful to have a link to the formal syntax, in a
   comment above the new atom
   constants. (https://tools.ietf.org/html/rfc3501#section-9)
5. DQUOTE seems to be inherited from
   https://tools.ietf.org/html/rfc2234 so maybe add a link to that too.
6. It would have been easier to review if the new constant names had
   exactly matched the names in the formal syntax eg ATOM_SPECIALS,
   LIST_WILDCARDS etc instead of _nonAtomChars
7. The comment "Perhaps turn the case where this actually removes
   bytes from the value into a warning and then an error, eventually."
   looks like it should be marked as TODO - in which case, add a
   ticket number also.
8. The unit test comments still contain links to RFC2822 which, as you
   noted in the ticket, are related to SMTP, not IMAP.

I think this should be merged once those RFC links have been updated.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [36904]) address review feedback

refs [#4392](https://github.com/twisted/twisted/issues/4392)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [36905]) Branching to 'imap4-keyword-atoms-4392-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set status to `closed`|
|-|-|

(In [36907]) Merge imap4-keyword-atoms-4392-2

Author: jothan, argonemyth, exarkun
Reviewer: exarkun, danmil, rwall
Fixes: [#4392](https://github.com/twisted/twisted/issues/4392)

Check for and filter out illegal bytes (given by application code) in the IMAP4
client's KEYWORD query support.
